contracts/contracts/token/WOETH.sol

```diff
@@ -188,5 +222,44 @@ contract WOETH is ERC4626, Governable, Initializable {
         uint256 creditsBefore = _getOETHCredits();
         oethAmount = super.redeem(woethAmount, receiver, owner);
         oethCreditsHighres -= creditsBefore - _getOETHCredits();
+        increaseYieldLimit();
+    }
+
+    function _transfer(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) internal override {
+        super._transfer(sender, recipient, amount);
+        increaseYieldLimit();
+    }
+
+    /**
+     * @dev This function safeguards rapid increases in WOETH -> OETH exchange ratio.
+     * Such an increase can be a result of a donation attack to the Vault contract.
+     */
+    function increaseYieldLimit() public {
+        // not yet time to increase the limit
+        if (block.timestamp < cptLimitEndTime) {
```

Comment:
I think there should be a way where we can cleanly allow pushing forward the end time with every action, rather than requiring one limit to end before the next begins. This would be much nicer user facing behavior. We do have to do some math to see how much of the partially completed phase to carry over.

Reply:
The current approach around limits is:
- very simple
- correctly handles the normal case use of the protocol
- correctly blocks donation attacks

But it is also feels like it has a lot weirdness when the limits are hit.
- 2x the limit is possible at the end of the period
- someone can get doubled yielded at the end of a period
- ratios can change mid block, separate from OETH ratios changing

Reply:
"someone can get doubled yielded at the end of a period"

Are you thinking that a person would be holding OETH, and knowing WOETH needs time to catch up, they would wrap their OETH just after a fat rebase?

"ratios can change mid block, separate from OETH ratios changing"

Do you think this can have negative consequences?

Reply:
So this is talking about behavior when the limits will be hit, and the time is past the end:

- User knows/causes a BIG rebase to be about to occur.
- They have/get a large OETH position.
- The call rebase(), collecting the OETH yield
- They then deposit into WEOTH, at a privileged exchange rate.
- The end of the deposit calls increaseYield. Limits are raised again
- User withdraws, collecting OETH at the new higher rate. No one else was able to do this.

Reply:
I'm actually concerned that is a full on attack, not just a yield steal. Minting funds at the wrong rate could be bad. This is the classic problem with the obvious path on rates based solutions.

Reply:
I think it's just a yield steal, since the someone can't mint at a lower rate than the existing users got in at.

Reply:
>> "ratios can change mid block, separate from OETH ratios changing"

> Do you think this can have negative consequences?

Yes, it allows someone to selectively take actions between these two. Presumably yield stealing.

Reply:
Ouch!

Confirmed with tests that this is a flashloanable enabled yield steal in the current implementation.

🟥 But it's actually much worse. Stuff does get really out of wack when things are at over the rate limit, and interacting users can actually end up with less funds than they started with depending on how they interact. It gets deeply squirrelly.